cicecore/cicedynB/dynamics/ice_dyn_eap.F90

```diff
@@ -392,6 +372,9 @@ subroutine eap (dt)
          call ice_HaloMask(halo_info_mask, halo_info, halomask)
       endif
 
+      ! velocities may have changed in dyn_prep2
+      call ice_HaloUpdate_vel(uvel, vvel, halo_info_mask)
```

Comment:
halo_info_mask is not defined unless maskhalo_dyn is true.  I find it a little awkward that halo_info_mask is passed in here and then in ice_haloUpdate_vel, halo_info_mask is only used if maskhalo_dyn is true, otherwise halo_info is used (which has been passed into the subroutine via a "use" statement).

I'm wondering if a clearer implementation might not be

```
if (maskhalo_dyn) then
  call ice_HaloUpdate_vel(uvel,vvel,halo_info_mask)
else
  call ice_HaloUpdate_vel(uvel,vvel,halo_info)
endif

```

and then remove the if check inside ice_HaloUpdate_vel.  Also, would it make sense to generalize the ice_HaloUpdate_vel so it looks more like

`call ice_HaloUpdate_vel(f1, f2, halo_info, field_loc)
`

which would make it more usable and more closely match the other HaloUpdate interfaces.  I'm also torn about whether this should be in ice_boundary or not.  If I had a choice, I'd put it in ice_boundary now and then create a separate issue that targeted improved reusability (or even merged) the mpi and serial comm directories.

Reply:
I agree this is a little awkward. Today I've been working on refactoring that, it's not as easy as it looks.

I tried just moving the subroutine to module **`ice_boundary`**, but that creates a circular dependency between **`ice_domain`** and **`ice_boundary`** (because I have to `use ice_domain` in the subroutine to get access to `halo_info`, `maskhalo_dyn`, and `nblocks`.

So I'll try a different approach tomorrow, keeping the subroutine in **`ice_dyn_shared`** but generalizing the interface a bit, as you suggest.

Reply:
I tried another approach in af8d03e. Since as I explained above moving the `ice_Haloupdate_vel` subroutine to **`ice_boundary`** does not appear feasible, I think that leaving it in **`ice_dyn_shared`** makes more sense. Then it is clear that it's only for the velocity, so `field_loc_NEcorner` stays hardcoded.

As I explain in the commit message of af8d03e, I chose to move the declaration of `halo_info_mask` to **`ice_dyn_shared`** as a module variable. I'm open to undoing that if we feel it's not something we want. It does simplify the interfaces in **`ice_dyn_vp`** since `halo_info_mask`  does not have to be passed down the call stack.

Reply:
I had a quick look at af8d03e.  I am pretty uncomfortable with halo_info_mask becoming reusable module data.  halo_info is set at initialization and is static.  halo_info_mask is very dynamic.  In fact, the correct version may vary in different parts of the code depending what you want to mask.  It certainly varies in time.  My preference continues to be that halo_info_mask be local, computed when needed locally, used locally, passed into other routines, and then destroyed.  There is too much risk someone is going to use/resuse that module data when it's going to contain stale data or nothing at all.  I also don't like the disconnect between computing halo_info_mask in one place and how delicately that interacts with the halo update in another part of the code.  

I continue to believe if we are going to want an ice_haloUpdate_vel, it should be more general (call ice_HaloUpdate_vel(uvel, vvel, halo_info, field_loc)) and in ice_boundary.F90.  If there is some circular logic issues, then maybe we need to think about moving some stuff out of ice_domain.F90 so it can be used in ice_boundary.F90.

I understand that aggregating uvel and vvel into a single variable might be part of the problem.  That's done for performance.  We can always haloUpdate each variable separately if we don't like the look of it.  All of this (creating the maskhalo and copying the two fields into one) is all done for performance.  I feel like the new implementation is creating more confusion/problems than it fixes.  I'm not sure what problem we're trying to fix.

The other thing is maybe we could create a method that creates "f2" (call copy2to1(f2,uvel,vvel)), then a call ice_haloUpdate as before with an if test on the maskhalo, then a copy1to2(f2,uvel,vvel) .  Is the problem that you don't like to have to copy in/out of f2 not in a reused subroutine. 

Having said all that, I think this can maybe be fixed by changing some names.  I think use of ice_HaloUpdate_vel and halo_info_mask are part of my problem.  Those are way too generic and are giving me heartburn.  If ice_HaloUpdate_vel were renamed to something like ice_dyn_velhalo and the module data, halo_info_mask were renamed to something like ice_dyn_velhalomask, I could handle this a lot better.  I still think ice_dyn_velhalomask should not be module data and prefer an implementation more like

```
if (maskhalo_dyn) then
  call ice_dyn_velhalo(uvel,vvel,halo_info_mask,field_loc_NEcorner)
else
  call ice_dyn_velhalo(uvel,vvel,halo_info,field_loc_NEcorner)
endif
```
which means we could keep the name halo_info_mask as it's no longer module data and just local.  We also get rid of the mask_halo_dyn check inside the "ice_HaloUpdate_vel" call.  

I guess in general, I still prefer to see a more generic interface created in this case.  If we want to continue down this path, I think at least we need to come up with names that don't feel out of place.

Reply:
> I am pretty uncomfortable with halo_info_mask becoming reusable module data.

I had a feeling you would not be comfortable with this change. In fact I was unsure myself if I liked my solution because as you say it makes it easy for callers of "ice_HaloUpdate_vel" to forget that the `halo_info_mask` has to be previously correctly created. I'll revert that change.

>  I feel like the new implementation is creating more confusion/problems than it fixes. I'm not sure what problem we're trying to fix.

My main goal was reducing code duplication.  

> I continue to believe if we are going to want an ice_haloUpdate_vel, it should be more general (call ice_HaloUpdate_vel(uvel, vvel, halo_info, field_loc)) and in ice_boundary.F90. If there is some circular logic issues, then maybe we need to think about moving some stuff out of ice_domain.F90 so it can be used in ice_boundary.F90.

OK. I'll try to work on that. The other thing that felt awkward to me when I tried this solution is that the `fld2` array used to combine uvel and vvel needs to be allocated somewhere (and I think for performance we wouldn't want to allocate it inside the subroutine each time it is called, but I could be wrong). If we move the subroutine to ice_boundary, then it makes more sense for `fld2` to be private module data in ice_boundary, but there is no existing `alloc_boundary` subroutine, so it would have to be added and all drivers updated to call it, which I wanted to avoid...

Reply:
I think there are a number of issues to get this refactoring right.  The allocation issue is a good point.  It's better if fld2 is local and we can reuse it during the loop.  I understand there could be a bit more reuse, and I'm supportive of that.  Sorry for being a pain, I just want to make sure the implementation is relatively clean if we do change it.

Reply:
Yes I understand. I'll come back to this the week after next.

Reply:
I re-immersed myself in this today and came to the conclusion that I a big refactor would involve spending more time on this than I want to/can invest. So what I did was reverting the changes I made (removing `ice_HaloUpdate_vel`) and simply introduce two new subroutines, `stack_velocity_field` and `unstack_velocity_field`, like you suggested above. 

There is still a lot of copied code but at least the amount of code reuse is increased a bit.

The changes are here: 6afd6f4.